token-swap/program/src/processor.rs

```diff
@@ -676,6 +706,272 @@ impl Processor {
+        let swap_info = next_account_info(account_info_iter)?;
+        let authority_info = next_account_info(account_info_iter)?;
+        let pool_mint_info = next_account_info(account_info_iter)?;
+        let source_info = next_account_info(account_info_iter)?;
+        let swap_token_a_info = next_account_info(account_info_iter)?;
+        let swap_token_b_info = next_account_info(account_info_iter)?;
+        let destination_info = next_account_info(account_info_iter)?;
+        let pool_fee_account_info = next_account_info(account_info_iter)?;
+        let token_program_info = next_account_info(account_info_iter)?;
+
+        let token_swap = SwapInfo::unpack(&swap_info.data.borrow())?;
+        let destination_account =
+            Self::unpack_token_account(destination_info, &token_swap.token_program_id)?;
+        let swap_token_a =
+            Self::unpack_token_account(swap_token_a_info, &token_swap.token_program_id)?;
+        let swap_token_b =
+            Self::unpack_token_account(swap_token_b_info, &token_swap.token_program_id)?;
+
+        let trade_direction = if destination_account.mint == swap_token_a.mint {
+            TradeDirection::AtoB
+        } else if destination_account.mint == swap_token_b.mint {
+            TradeDirection::BtoA
+        } else {
+            return Err(SwapError::IncorrectSwapAccount.into());
+        };
+
+        let (destination_a_info, destination_b_info) = match trade_direction {
+            TradeDirection::AtoB => (Some(destination_info), None),
+            TradeDirection::BtoA => (None, Some(destination_info)),
+        };
+        Self::check_accounts(
+            &token_swap,
+            program_id,
+            swap_info,
+            authority_info,
+            swap_token_a_info,
+            swap_token_b_info,
+            pool_mint_info,
+            token_program_info,
+            destination_a_info,
+            destination_b_info,
+            Some(pool_fee_account_info),
+        )?;
+
+        let pool_mint = Self::unpack_mint(pool_mint_info, &token_swap.token_program_id)?;
+        let pool_mint_supply = to_u128(pool_mint.supply)?;
+        let (swap_token_a_amount, swap_token_b_amount) = match trade_direction {
+            TradeDirection::AtoB => (
+                to_u128(
+                    swap_token_a
+                        .amount
+                        .checked_sub(destination_token_amount)
+                        .ok_or(SwapError::CalculationFailure)?,
+                )?,
+                to_u128(swap_token_b.amount)?,
+            ),
+            TradeDirection::BtoA => (
+                to_u128(swap_token_a.amount)?,
+                to_u128(
+                    swap_token_b
+                        .amount
+                        .checked_sub(destination_token_amount)
+                        .ok_or(SwapError::CalculationFailure)?,
+                )?,
+            ),
+        };
+
+        let burn_pool_token_amount = token_swap
```

Comment:
To do exact amount out, we have to calculate the amount and then tack the fee on afterward